nlprule/src/rule/grammar.rs

```diff
@@ -176,18 +163,18 @@ impl Synthesizer {
         let make_uppercase = !starts_with_conversion
             && graph.groups()[graph.get_index(start)..]
                 .iter()
-                .find_map(|x| x.tokens(graph.tokens()).next())
+                .find_map(|x| x.tokens(sentence).next())
                 .map(|first_token| {
                     (self.use_titlecase_adjust
                         && first_token
-                            .word
+                            .word()
                             .text
                             .as_ref()
                             .chars()
                             .next()
                             .expect("token must have at least one char")
```

Comment:
I'd recommend to be defensive about these guarantees which are not in the scope of the same `fn` and rather use a `filter_map`.

Also consider formulating the expects as the invariant that you expect (harr) to hold in a non-negating manner. It's not an user facing error after all :) so here. "Zero length tokens are filtered by the caller."

Reply:
Good point, I'll fix that. I think a `fn first(&self) -> Token` and `fn first(&self) -> char` on the `Sentence` / `Word` would make sense to to check that expectation on the struct itself, but I'll keep that in mind for a future PR, this is large enough as it is.